python/rmm/_lib/device_buffer.pyx

```diff
@@ -34,6 +34,17 @@ cdef class DeviceBuffer:
     def size(self):
         return self.c_obj.get()[0].size()
 
+    @property
+    def __cuda_array_interface__(self):
+        cdef dict intf = {
+            "data": self.ptr,
+            "shape": (self.size,),
```

Comment:
Does this work as expected given we have both a Python `size` property and a Cython `size` method?

Reply:
In exploring this question, I'm not actually sure the current implementation works. To underline this, I can't compile this example.

```cython
cdef class Test:
    cdef size_t _size

    def __init__(self, size):
        self._size = size

    @property
    def size(self):
        return self._size

    cdef size_t size(self):
        return self._size
```

I think we have to choose whether we want properties or `cdef`s in this extension type. So which one do we want?

Reply:
@jakirkham is the above because of using the name `size` twice or just the combination of properties and cdefs?

Reply:
Where? The attribute has an underscore. The constructor input is also not a member of `self`. So I don't think it matters.

That said, if I try prefixing both to disambiguate them further (like below) the compilation error still occurs.

```cython
cdef class Test:
    cdef size_t _my_size

    def __init__(self, new_size):
        self._my_size = new_size

    @property
    def size(self):
        return self._my_size

    cdef size_t size(self):
        return self._my_size
```

I think it is just the property and `cdef` function duplicates, but please let me know if I'm missing something else.

Reply:
Sorry I misunderstood. What you posted above makes sense. It seems like we should disambiguate the Python vs C methods and attributes.

Reply:
No worries. There were a lot of different uses of `size`. So it didn't help. Apologies for not being clearer initially.

Ok, here are a few thoughts on possible directions. What do you think?

1. Prefix the C implementations (with something like `c_` or even `_`)
2. Prefix the Python implementations (with something like `py_`)
3. Copy this info into `readonly` attributes that both C and Python could use
4. Drop the C implementations (if they are not needed)

Additionally if we go with something like 1 or 2. It would be good to have the Python implementations call the C implementations to avoid code duplication.

Reply:
Given this seems to just be `size` that is of issue, we could also consider using `__len__` for the Python side.

Reply:
I've raised issue ( https://github.com/rapidsai/rmm/issues/178 ) given this is a bit orthogonal from the work here. Happy to work with you to solve it though. 🙂